lib/IRGen/GenStruct.cpp

```diff
@@ -329,6 +332,146 @@ namespace {
     }
   };
 
+  // A common base class for loadable and address-only non-trivially copyable
+  // C structs (i.e., they contain ARC fields).
+  template <class Impl, class Base, class FieldInfoType = ClangFieldInfo>
+  class ClangNonTrivialRecordTypeInfoBase :
+     public StructTypeInfoBase<Impl, Base, FieldInfoType> {
+    using super = StructTypeInfoBase<Impl, Base, FieldInfoType>;
+
+    const clang::RecordDecl *ClangDecl;
+  protected:
+    template <class... As>
+    ClangNonTrivialRecordTypeInfoBase(StructTypeInfoKind kind,
+                                      const clang::RecordDecl *clangDecl,
+                                      As &&...args)
+      : super(kind, std::forward<As>(args)...), ClangDecl(clangDecl)
+    {}
+
+    const clang::RecordDecl *getClangDecl() const { return ClangDecl; }
+
+  public:
+    void initializeWithCopy(IRGenFunction &IGF, Address destAddr,
+                            Address srcAddr, SILType T, bool isOutlined) const {
+      clang::CodeGen::callNonTrivialCStructCopyConstructor(
```

Comment:
For cross-referencing purposes, these function calls have been added in https://reviews.llvm.org/D60161 (pending review).